lib/active_merchant/billing/gateways/checkout_v2.rb

```diff
@@ -178,13 +184,16 @@ def add_customer_data(post, options)
         post[:payment_ip] = options[:ip] if options[:ip]
         address = options[:billing_address]
         if address
-          post[:billing_address] = {}
-          post[:billing_address][:address_line1] = address[:address1] unless address[:address1].blank?
-          post[:billing_address][:address_line2] = address[:address2] unless address[:address2].blank?
-          post[:billing_address][:city] = address[:city] unless address[:city].blank?
-          post[:billing_address][:state] = address[:state] unless address[:state].blank?
-          post[:billing_address][:country] = address[:country] unless address[:country].blank?
-          post[:billing_address][:zip] = address[:zip] unless address[:zip].blank?
+          post[:customer][:name] = options[:billing_address][:name]
+          billing_address = {}
+          billing_address[:billing_address] = {}
+          billing_address[:billing_address][:address_line1] = address[:address1] unless address[:address1].blank?
+          billing_address[:billing_address][:address_line2] = address[:address2] unless address[:address2].blank?
+          billing_address[:billing_address][:city] = address[:city] unless address[:city].blank?
+          billing_address[:billing_address][:state] = address[:state] unless address[:state].blank?
+          billing_address[:billing_address][:country] = address[:country] unless address[:country].blank?
+          billing_address[:billing_address][:zip] = address[:zip] unless address[:zip].blank?
+          post[:source] ? post[:source].merge!(billing_address) : post.merge!(billing_address)
```

Comment:
It looks like the `billing_address` is empty {}. Do you mean `billing_address[:billing_address]` ?

Reply:
The `billing_address` is set as an empty hash, then we assign all data to it. 
In the end, it is merged in the `post` hash:
```ruby
post[:source] ? post[:source].merge!(billing_address) : post.merge!(billing_address)
```